### PR TITLE
DEVELOP-1437: Test data and automatic tests

### PR DESCRIPTION
**Pull request by matrulda**
This PR adds test data and automatic tests for the pipeline. 

In summary this has been done:
- The folder `test_data` has been added with files to be able to run a complete integration test of of the pipeline. Among other things, it contains a run folder (210510_M03910_0104_000000000-JHGJL) from Illumina's public Demo space. The majority of files in this PR originates from this folder, they do not need a detailed inspection.
- The complete integration test can be run using the new `test` profile. It is used like this: 
```
nextflow run main.nf -profile dev,test,singularity
```
- The folder `tests` contains:
  - Integration tests: In these tests the pipeline is run with the `test` profile and then the output is validated. Validation of the output includes verifying that reports exist and that they contain the sections they should. 
  - Unit tests: These are unit tests for the python scripts in `bin` that we use in the pipeline.
- A Github Actions workflow has been added so that tests are run every time code is pushed to the repo.
- I have formatted all python code with [black](https://github.com/psf/black) and it is now enforced through a check in the test workflow. 
- I've added instructions on how to perform tests locally.
- Path to images has been updated so that we use images built for Singularity instead of Docker. Singularity can handle converting Docker images, but this became an issue when running the tests on Github's servers. Using these images makes the pipeline faster and more efficient. Related to this change I had to use new versions of software we use to be able to find publicly hosted images. I've made the assessment that the upgrades (minor and patch) are safe. 
- Configs in `config` has been sorted into subfolders `nextflow_config` and  `tools_config`.

### CODE REVIEW

File: test_data/test_config/fastq_screen.conf [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+###########
+## Bowtie #
+###########
+## If the bowtie binary is not in your PATH then you can
+## set this value to tell the program where to find it.
+## Uncomment the line below and set the appropriate location
+##
+
+BOWTIE2 /usr/local/bin/bowtie2
```
Comment (GitEdvard) [inline test_data/test_config/fastq_screen.conf:9]: The text above indicates that this line should be commented out by default. Should it be?

File: requirements-dev.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+pytest==6.2.5
+black==21.11b1
+beautifulsoup4==4.10.0
+checkqc==3.6.6
+lxml==4.6.4
```
Comment (GitEdvard) [inline requirements-dev.txt:5]: I wonder, should nextflow be added to this list?

File: README.md [hunk 3]
```diff
@@ -52,6 +52,27 @@ There are two primary branches of this project:
 - `master`: The stable release branch
 - `dev`: The development and test branch, to which pull requests should be made.
 
-### Known issues:
+### Running tests locally
```
Comment (GitEdvard) [inline README.md]: (This is somewhat unrelated to the specific changes here) I wonder what dependencies are involved in the integration tests? I guess it is 1) fastq-screen db, 2) nextflow, 3) singularity. Have I missed anything? How about this slurm (?) we talked about the other day?

File: tests/integration_tests/test_validate_output.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        "general_stats",
+        "rrna",
+        "sequencing_metadata",
+        "fastq_screen",
+        "fastqc",
+    ]
+    return sections
+
+
+def test_results_dirs_exist(flowcell_report_dir, project_reports_dir):
```
Comment (GitEdvard) [inline tests/integration_tests/test_validate_output.py]: The assertions here is kind of hanging in the air. How does it come that, for example, flowcell_report_dir exists without any explicit action? Generally, I like tests to have extra verboseness, with an explicit "act" session. Perhaps you could add a "ensure_nextflow_has_run()" call or the like?

File: tests/integration_tests/test_validate_output.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+import pytest
+import os.path
+import subprocess
+from bs4 import BeautifulSoup
+import itertools
+
+
+@pytest.fixture(scope="session", autouse=True)
+def result_dir(tmpdir_factory):
```
Comment (GitEdvard) [inline tests/integration_tests/test_validate_output.py]: When naming a function to a noun, I anticipate a rather simple fetch algorithm. But here, the main action for the test is kind of hidden here, namely to start a nextflow pipeline. I would like a separate and explicit call for this.

File: tests/integration_tests/test_validate_output.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    yield result_dir
+
+
+@pytest.fixture
+def flowcell_report_dir(result_dir):
+    return os.path.join(result_dir, "flowcell_report")
+
+
+@pytest.fixture
+def project_reports_dir(result_dir):
```
Comment (GitEdvard) [inline tests/integration_tests/test_validate_output.py]: I wonder if "reports" belong in this variable name? I can't find reports within this search path.

File: tests/integration_tests/test_validate_output.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    return sections
+
+
+def test_results_dirs_exist(flowcell_report_dir, project_reports_dir):
+    assert os.path.isdir(flowcell_report_dir)
+    assert os.path.isdir(project_reports_dir)
+
+
+def test_project_dirs_exist(project_reports_dir, projects):
+    for project in projects:
```
Comment (GitEdvard) [inline tests/integration_tests/test_validate_output.py:112]: When reading tests like these, I would like to get an easy available picture of the folder tree generated by the pipeline. I find it a bit cumbersome that the folder paths are parameterized, and hence fragmented. In order to get this picture of the folder structure, I have to assemble them in a notebook beside. The general advise I've got for automatic tests are to favor readability and verboseness over the DRY principle. I wonder if you could consider write them out as they are in each test? Like "for project in ['duck', 'wolf']", and project_reports_dir = r'path/to/reports'.

File: tests/integration_tests/test_validate_output.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+import os.path
+import subprocess
+from bs4 import BeautifulSoup
+import itertools
+
+
+@pytest.fixture(scope="session", autouse=True)
+def result_dir(tmpdir_factory):
+    result_dir = tmpdir_factory.mktemp("results")
+    subprocess.run(
```
Comment (GitEdvard) [inline tests/integration_tests/test_validate_output.py:14]: I wonder if the directory for the previous run should be wiped before running the pipeline? What happens if you 1) run integrations test, 2) make some changes, 3) run integration tests again? Is there any risk that a file or folder belonging to the previous run is affecting the second run on integration test?

**Review by GitEdvard — COMMENTED** [review]
Good work! I appreciate that the commits were relative restrictive in size and had a good explanation. As you can see, I "dived" in to the structure of the tests a bit. Let's see what you think about that! Beside from that, I think it's important to look over if the test directory is wiped between runs or not. (if it's not, I think you should implement  it)

Edit: I've checked now, test directory is handled by the tmp-factory. You don't need to do anything.

File: test_data/test_config/fastq_screen.conf (hunk 1 above)
Comment (matrulda) [reply test_data/test_config/fastq_screen.conf:9]: Ah, I think that comment is a leftover from the example config. That is the correct path to bowtie2 in the fastq_screen container. I will remove the comment.

File: requirements-dev.txt (hunk 2 above)
Comment (matrulda) [reply requirements-dev.txt:5]: If we used conda to set this up we could, since it can be downloaded via bioconda. To keep things simple, the instructions only includes how to set it up with virtualenv. It is expected that pre-requisites are installed to be able to run the integration tests.

File: README.md (hunk 3 above)
Comment (matrulda) [reply README.md]: Only nextflow and singularity are needed, a minimal fastq_screen db is included in the test_data set. I can add info about that to the README.  👍 

Good point, there are no tests for the irma profile (where slurm is used). I think it would be quite complex to set that up. I think it is fine that we don't include tests for that since using slurm is a nextflow feature more than a feature of this pipeline.

File: tests/integration_tests/test_validate_output.py (hunk 8 above)
Comment (matrulda) [reply tests/integration_tests/test_validate_output.py:14]: The tmpdir_factory makes a new folder every test session, so it should be ok.

File: tests/integration_tests/test_validate_output.py (hunk 7 above)
Comment (matrulda) [reply tests/integration_tests/test_validate_output.py:112]: Good point! I've tried to make it more readable in https://github.com/Molmed/seqreports/pull/46/commits/58d7a4b7844062fc2f63f50b79be8aef771167fa . I think it better follows the Arrange, Act, Assert model now. Even though the "Act" part comes from `result_dir` as input. I'd like to keep it this way to make use of the pytest fixture feature, but let me know if you still think it's too unclear.

File: test_data/test_config/fastq_screen.conf (hunk 1 above)
Comment (matrulda) [reply test_data/test_config/fastq_screen.conf:9]: Fixed in https://github.com/Molmed/seqreports/pull/46/commits/75e3d57f0a3c06593d3d6ec506c6fdb1b4ed3b47

File: requirements-dev.txt (hunk 2 above)
Comment (matrulda) [reply requirements-dev.txt:5]: Fixed in https://github.com/Molmed/seqreports/pull/46/commits/5d508dde7009ebd22492d0c41599a61fdfa5c0bd

File: README.md (hunk 3 above)
Comment (matrulda) [reply README.md]: Fixed in https://github.com/Molmed/seqreports/pull/46/commits/5d508dde7009ebd22492d0c41599a61fdfa5c0bd

File: tests/integration_tests/test_validate_output.py (hunk 4 above)
Comment (matrulda) [reply tests/integration_tests/test_validate_output.py]: See https://github.com/Molmed/seqreports/pull/46#discussion_r798716428

File: tests/integration_tests/test_validate_output.py (hunk 5 above)
Comment (matrulda) [reply tests/integration_tests/test_validate_output.py]: Good point. I tried to address this in https://github.com/Molmed/seqreports/pull/46/commits/58d7a4b7844062fc2f63f50b79be8aef771167fa

File: tests/integration_tests/test_validate_output.py (hunk 6 above)
Comment (matrulda) [reply tests/integration_tests/test_validate_output.py]: Fixed in https://github.com/Molmed/seqreports/pull/46/commits/58d7a4b7844062fc2f63f50b79be8aef771167fa

**Comment by matrulda** [comment]
@GitEdvard Thanks for the comments! I've addressed them, let me know if you think it's ok.

**Comment by GitEdvard** [comment]
Great!
